onchain/rollups/test/foundry/consensus/quorum/QuorumFactory.t.sol

```diff
@@ -0,0 +1,161 @@
+            _salt
+        );
+
+        emit QuorumCreated(quorumValidators, quorum);
+
+        // Precalculated address must match actual address
+        assertEq(precalculatedAddress, address(quorum));
+        checkFactoryLogs(quorumValidators, quorum);
+    }
+
+    function testAlreadyDeployedNewQuorumDeterministic(
+        uint256 _numValidators,
+        bytes32 _salt
+    ) public {
+        vm.assume(_numValidators > 1);
+        vm.assume(_numValidators < 50);
+
+        address[] memory quorumValidators = generateValidators(_numValidators);
+        uint256[] memory shares = generateShares(quorumValidators);
+
+        IHistory history = new History(msg.sender);
+
+        factory.newQuorum(quorumValidators, shares, history, _salt);
+
+        //Deploy already deployed quorum
+        vm.expectRevert();
+        factory.newQuorum(quorumValidators, shares, history, _salt);
+    }
+
+    // HELPER FUNCTIONS
+    function generateValidators(
+        uint256 _numValidators
+    ) internal pure returns (address[] memory) {
+        address[] memory validators = new address[](_numValidators);
+        for (uint256 i = 0; i < _numValidators; i++) {
+            validators[i] = vm.addr(i + 1);
+        }
+        return validators;
+    }
+
+    function generateShares(
+        address[] memory validators
+    ) internal pure returns (uint256[] memory) {
+        //generate a random number of shares for each validator
+        uint256[] memory shares = new uint256[](validators.length);
+        for (uint256 i; i < shares.length; ++i) {
+            uint256 share = uint256(
+                keccak256(abi.encodePacked(i, validators[i]))
+            ) % 100;
+            shares[i] = (share > 0) ? share : ++share;
+        }
+        return shares;
+    }
+
+    function checkFactoryLogs(
+        address[] memory _quorumValidators,
+        Quorum quorum
+    ) internal {
+        Vm.Log[] memory entries = vm.getRecordedLogs();
+        uint256 numOfQuorumsCreated;
+
+        for (uint256 i = 0; i < entries.length; i++) {
+            Vm.Log memory entry = entries[i];
+
+            if (
+                entry.topics[0] == QuorumCreated.selector &&
+                entry.emitter == address(factory)
+            ) {
+                ++numOfQuorumsCreated;
+                QuorumCreatedEvent memory quorumEvent;
+                (quorumEvent.quorumValidators, quorumEvent.quorum) = abi.decode(
+                    entry.data,
+                    (address[], Quorum)
+                );
+
+                //Check Validators length in decoded data and each validator address
+                checkEq0(
+                    abi.encodePacked(quorumEvent.quorumValidators),
+                    abi.encodePacked(_quorumValidators)
+                );
+
+                //Compare quorum address
+                assertEq(address(quorum), address(quorumEvent.quorum));
+                assertEq(numOfQuorumsCreated, 1);
+            }
+        }
```

Comment:
We should check that `numOfQuorumsCreated` is `1` after the loop.

Reply:
Not really a big difference doing it before the previous assert. But I'll change it.